firmware/stm32f0/Core/Src/adc-light-sens.c

```diff
@@ -0,0 +1,53 @@
+/*
+ * adc-light-sens.c
+ *
+ *  Created on: Jan 13, 2024
+ *      Author: Matt L.
+ */
+
+
+#include "adc-light-sens.h"
+
+#define ADC_BUF_LENGTH  50
+#define MAX_LIGHT_LEVEL 4095    // Max value on 12-bit ADC
+#define ALPHA           90      // Alpha value for moving avg. (out of 100)
+#define TRANSITION_BUFF 204
+
+ADC_HandleTypeDef *adc;
+static uint16_t light_threshold;
+static uint16_t adc_buffer[ADC_BUF_LENGTH];
+static uint32_t sample_avg;
+static uint32_t moving_avg;
+static bool is_dark_room;
+
+void LightSens_Init(ADC_HandleTypeDef *hadc, uint16_t threshold)
+{
+    adc = hadc;
+    is_dark_room = 0;
+    light_threshold = threshold;
+    moving_avg = MAX_LIGHT_LEVEL;
+    HAL_ADC_Start_DMA(adc, (uint32_t*)adc_buffer, ADC_BUF_LENGTH);

```

Comment:
Looks like adc_buffer type cast on the HAL_ADC_Start_DMA call should be (uint16_t*) instead of (uint32_t*). Just based off the fact that adc_buffer is initialized above as uint16_t array.

Reply:
Yeah it looks kinda wrong lol. It's a pretty complicated explanation so I can go over it more in the meeting. Basically, the HAL_ADC_Start_DMA function can only accept a uint32_t pointer as an argument because the DMA stores everything in that type. But when the ADC takes samples the max. resolution is 12-bits so it reads it as uint16_t. When the ADC stores the data in the DMA it puts two uint16_t sample readings into each uint32_t DMA word. The DMA thinks these are uint32_t and will write to my adc_buffer as if it also stores uint32_t (because of the type cast when I pass the pointer parameter). But I can still access adc_buffer and read it as uint16_t because of how the ADC wrote the data in the first place